components/dashboard/src/workspaces/PersonalizedContent.tsx

```diff
@@ -0,0 +1,260 @@
+        title: "Building a business case for Gitpod",
+        label: "cde-roi-calculator",
+        priority: 8,
+        recommended: {
+            jobRole: ["enabling", "team-lead"],
+            explorationReasons: ["replace-remote-dev"],
+            signupGoals: ["efficiency-collab", "security"],
+        },
+    },
+    {
+        url: "https://www.gitpod.io/blog/whats-a-cloud-development-environment",
+        title: "What's a CDE",
+        label: "what-is-cde",
+        priority: 9,
+        recommended: {
+            jobRole: ["enabling", "team-lead"],
+        },
+    },
+];
+
+const defaultContent: ContentItem[] = [
+    {
+        url: "https://www.gitpod.io/blog/whats-a-cloud-development-environment",
+        title: "What's a CDE",
+        label: "what-is-cde",
+    },
+    {
+        url: "https://www.gitpod.io/solutions/onboarding",
+        title: "Onboarding developers in one click",
+        label: "onboarding-solutions",
+    },
+    {
+        url: "https://www.gitpod.io/blog/using-a-cde-roi-calculator",
+        title: "Building a business case for Gitpod",
+        label: "cde-roi-calculator",
+    },
+];
+
+const PersonalizedContent: React.FC = () => {
+    const user = useCurrentUser();
+    const [selectedContent, setSelectedContent] = useState<ContentItem[]>([]);
+
+    useEffect(() => {
+        if (!storageAvailable("localStorage")) {
+            // Handle the case where localStorage is not available
+            setSelectedContent(getFirstWeekContent(user));
+            return;
+        }
+
+        let content: ContentItem[] = [];
+        let lastShownContent: string[] = [];
+
+        try {
+            const storedContentData = localStorage.getItem("personalized-content-data");
+            const currentTime = new Date().getTime();
+
+            if (storedContentData) {
+                const { lastTime, lastContent } = JSON.parse(storedContentData);
+                const WEEK_IN_MS = 7 * 24 * 60 * 60 * 1000;
+                const weeksPassed = Math.floor((currentTime - lastTime) / WEEK_IN_MS);
+                lastShownContent = lastContent || [];
+
+                if (weeksPassed >= 1) {
+                    content = getRandomContent(contentList, 3, lastShownContent);
+                } else {
+                    content = getFirstWeekContent(user);
+                }
+            } else {
+                content = getFirstWeekContent(user);
+            }
+
+            localStorage.setItem(
+                "personalized-content-data",
+                JSON.stringify({
+                    lastContent: content.map((item) => item.label),
+                    lastTime: currentTime,
+                }),
+            );
+
+            setSelectedContent(content);
+        } catch (error) {
+            console.error("Error handling personalized content: ", error);
+            setSelectedContent(getRandomContent(contentList, 3, []));
+        }
+    }, [user]);
```

Comment:
Following [You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect), we might not :). `selectedContent` is only updated here in this effect, and we only rely on the user object for decisions. This could be a memo, also helping us to get rid of state. 

Here's my try at this:

```suggestion

    const selectedContent = useMemo<ContentItem[]>(() => {
        if (!storageAvailable("localStorage")) {
            // Handle the case where localStorage is not available
            return getFirstWeekContent(user);
        }

        let content: ContentItem[] = [];
        let lastShownContent: string[] = [];

        try {
            const storedContentData = localStorage.getItem("personalized-content-data");
            const currentTime = new Date().getTime();

            if (storedContentData) {
                const { lastTime, lastContent } = JSON.parse(storedContentData);
                const WEEK_IN_MS = 7 * 24 * 60 * 60 * 1000;
                const weeksPassed = Math.floor((currentTime - lastTime) / WEEK_IN_MS);
                lastShownContent = lastContent || [];

                if (weeksPassed >= 1) {
                    content = getRandomContent(contentList, 3, lastShownContent);
                } else {
                    content = getFirstWeekContent(user);
                }
            } else {
                content = getFirstWeekContent(user);
            }

            localStorage.setItem(
                "personalized-content-data",
                JSON.stringify({
                    lastContent: content.map((item) => item.label),
                    lastTime: currentTime,
                }),
            );

            return content;
        } catch (error) {
            console.error("Error handling personalized content: ", error);
            return getRandomContent(contentList, 3, []);
        }
    }, [user]);
```

Reply:
Note: this will also require updating imports, so it won't compile by just accepting the suggestion 🙏.

Reply:
This doesn't seem to work!!!

You can test by manually updating the time to 2 weeks (ago), and see if the content (on the visuals side) is not getting updated.

Reply:
We would need this to re-trigger once content will be updated (after 1 week or so):


<img width="1505" alt="image" src="https://github.com/user-attachments/assets/8df38f3f-481b-4b8d-bc3a-cdb0f583937c">


I've also tried to add key to force re-render when content changes

Reply:
I'll leave this as it is for now.